Split-brain when redis master restarted without restart of its resec

Hello

I found a reproducible situation: when you stop redis process on master, it becomes red in consul, but resec do not removes it from redis-master service.
Then after redis process is restarted, healthcheck sets it to green, and we now have double master setup, with slave connected to one of them.
Restart of resec process on former master resolves issue.

resec's log output from this master:

```
2018/03/19 17:05:42 [ERROR] Can't connect to redis running on 127.0.0.1:6379
2018/03/19 17:05:42 [INFO] Redis HealthCheck changed to NOT healthy
2018/03/19 17:05:42 [INFO] Received update for master from consul
2018/03/19 17:05:42 [INFO] No redis master services in Consul
2018/03/19 17:05:42 [INFO] Received update for master from consul
2018/03/19 17:05:42 [INFO] Redis master updated in Consul
2018/03/19 17:05:42 [INFO] Received update for master from consul
2018/03/19 17:05:42 [INFO] No redis master services in Consul
2018/03/19 17:05:47 [INFO] Received update for master from consul
2018/03/19 17:05:47 [ERROR] Can't connect to redis running on 127.0.0.1:6379
2018/03/19 17:05:52 [INFO] Redis HealthCheck changed to healthy
2018/03/19 17:05:52 [INFO] Trying to acquire leader lock
2018/03/19 17:05:52 [INFO] Received update for master from consul
2018/03/19 17:05:52 [ERROR] Found more than one master registered in Consul

```

--- a/init.go
+++ b/init.go
@@ -5,8 +5,10 @@ import (
 	"log"
 	"net/http"
 	"os"
+	"os/signal"
 	"strconv"
 	"strings"
+	"syscall"
 	"time"
 
 	"github.com/go-redis/redis"
@@ -103,6 +105,8 @@ type resec struct {
 	consulMasterServiceCh chan []*consulapi.ServiceEntry // channel where consul updates to the master service will be published
 	redis                 *redisConnection               // internal Redis state
 	redisReplicationCh    chan *redisHealth              // channel where redis updates will be published
+	sigCh                 chan os.Signal                 // signal handler channel
+	stopCh                chan interface{}               // stop channel for non-signal termination
 }
 
 // setup returns the default configuration for the ReSeC
@@ -133,7 +137,11 @@ func setup() (*resec, error) {
 		healthCheckTimeout:    2 * time.Minute,
 		consulMasterServiceCh: make(chan []*consulapi.ServiceEntry, 1),
 		redisReplicationCh:    make(chan *redisHealth, 1),
+		sigCh:                 make(chan os.Signal, 1),
+		stopCh:                make(chan interface{}, 1),
 	}
+
+	signal.Notify(config.sigCh, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT)
 
 	if logLevel := os.Getenv(LogLevel); logLevel != "" {
 		config.logLevel = logLevel

--- a/main.go
+++ b/main.go
@@ -17,10 +17,5 @@ func main() {
 	go resec.watchRedisReplicationStatus()
 	go resec.watchConsulMasterService()
 
-	defer func() {
-		log.Printf("[INFO] Shutting down ...")
-		resec.stop()
-	}()
-
-	resec.start()
+	resec.run()
 }

--- a/main.go
+++ b/main.go
@@ -12,6 +12,7 @@ func main() {
 		log.Fatal(err)
 	}
 
+	resec.waitForRedisToBeReady()
 	go resec.watchRedisReplicationStatus()
 	go resec.watchConsulMasterService()
 

--- a/main.go
+++ b/main.go
@@ -13,6 +13,7 @@ func main() {
 	}
 
 	resec.waitForRedisToBeReady()
+	go resec.watchRedisUptime()
 	go resec.watchRedisReplicationStatus()
 	go resec.watchConsulMasterService()
 

--- a/redis.go
+++ b/redis.go
@@ -43,26 +43,58 @@ func (rc *resec) runAsMaster() error {
 
 // watchRedisReplicationStatus checks redis replication status
 func (rc *resec) watchRedisReplicationStatus() {
-	timer := time.NewTimer(rc.healthCheckInterval)
+	ticker := time.NewTicker(rc.healthCheckInterval)
 
-	for {
-		select {
-		case <-timer.C:
-			log.Println("[DEBUG] Checking redis replication status")
+	for ; true; <-ticker.C {
+		log.Println("[DEBUG] Checking redis replication status")
 
-			result, err := rc.redis.client.Info("replication").Result()
-			if err != nil {
-				log.Printf("[ERROR] Can't connect to redis running on %s", rc.redis.address)
-				result = fmt.Sprintf("Can't connect to redis running on %s", rc.redis.address)
-			}
-
-			rc.redisReplicationCh <- &redisHealth{
-				output:  result,
-				healthy: err == nil,
-			}
-
-			timer.Reset(rc.healthCheckInterval)
+		result, err := rc.redis.client.Info("replication").Result()
+		if err != nil {
+			log.Printf("[ERROR] Can't connect to redis running on %s", rc.redis.address)
+			result = fmt.Sprintf("Can't connect to redis running on %s", rc.redis.address)
 		}
+
+		rc.redisReplicationCh <- &redisHealth{
+			output:  result,
+			healthy: err == nil,
+		}
+	}
+}
+
+// watchRedisUptime checks redis server uptime
+func (rc *resec) watchRedisUptime() {
+	lastUptime := 0
+
+	ticker := time.NewTicker(rc.healthCheckInterval)
+	for ; true; <-ticker.C {
+		log.Println("[DEBUG] Checking redis server info")
+
+		result, err := rc.redis.client.Info("server").Result()
+		if err != nil {
+			log.Printf("[ERROR] Could not query for server info: %s", err)
+			continue
+		}
+
+		parsed := parseKeyValue(result)
+		uptimeString, ok := parsed["uptime_in_seconds"]
+		if !ok {
+			log.Printf("[ERROR] Could not find 'uptime_in_seconds' in server info respone")
+			continue
+		}
+
+		uptime, err := strconv.Atoi(uptimeString)
+		if err != nil {
+			log.Printf("[ERROR] Could not parse 'uptime_in_seconds' to integer")
+			continue
+		}
+
+		if uptime < lastUptime {
+			log.Printf("[ERROR] Current uptime (%d) is less than previous (%d) - Redis likely restarted - stopping resec", uptime, lastUptime)
+			rc.stopCh <- true
+			continue
+		}
+
+		lastUptime = uptime
 	}
 }
 

--- a/redis.go
+++ b/redis.go
@@ -64,6 +64,8 @@ func (rc *resec) watchRedisReplicationStatus() {
 // watchRedisUptime checks redis server uptime
 func (rc *resec) watchRedisUptime() {
 	lastUptime := 0
+	connectionErrors := 0
+	allowedConnectionErrors := 3
 
 	ticker := time.NewTicker(rc.healthCheckInterval)
 	for ; true; <-ticker.C {
@@ -71,9 +73,17 @@ func (rc *resec) watchRedisUptime() {
 
 		result, err := rc.redis.client.Info("server").Result()
 		if err != nil {
-			log.Printf("[ERROR] Could not query for server info: %s", err)
+			log.Printf("[WARN] Could not query for server info: %s", err)
+			connectionErrors++
+
+			if connectionErrors > allowedConnectionErrors {
+				log.Printf("[ERROR] Too many connection errors, shutting down")
+				rc.stopCh <- true
+			}
+
 			continue
 		}
+		connectionErrors = 0
 
 		parsed := parseKeyValue(result)
 		uptimeString, ok := parsed["uptime_in_seconds"]
@@ -117,7 +127,7 @@ func (rc *resec) waitForRedisToBeReady() {
 				continue
 			}
 
-			if loading == "1" {
+			if loading != "0" {
 				log.Printf("[INFO] Redis is not ready yet, currently loading data from disk")
 				continue
 			}

--- a/redis.go
+++ b/redis.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"log"
 	"strconv"
+	"strings"
 	"time"
 )
 
@@ -63,4 +64,54 @@ func (rc *resec) watchRedisReplicationStatus() {
 			timer.Reset(rc.healthCheckInterval)
 		}
 	}
+}
+
+func (rc *resec) waitForRedisToBeReady() {
+	t := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-t.C:
+			persistenceString, err := rc.redis.client.Info("persistence").Result()
+			if err != nil {
+				log.Printf("[ERROR] could not query for redis persistence info: %s", err)
+				continue
+			}
+
+			persistence := parseKeyValue(persistenceString)
+			loading, ok := persistence["loading"]
+			if !ok {
+				log.Printf("[ERROR] could not find 'persistence.loading' key")
+				continue
+			}
+
+			if loading == "1" {
+				log.Printf("[INFO] Redis is not ready yet, currently loading data from disk")
+				continue
+			}
+
+			log.Printf("[INFO] Redis is ready to serve traffic")
+			return
+		}
+	}
+}
+
+func parseKeyValue(str string) map[string]string {
+	res := make(map[string]string)
+
+	lines := strings.Split(str, "\r\n")
+	for _, line := range lines {
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+
+		pair := strings.Split(line, ":")
+		if len(pair) != 2 {
+			continue
+		}
+
+		res[pair[0]] = pair[1]
+	}
+
+	return res
 }

--- a/resec.go
+++ b/resec.go
@@ -7,7 +7,9 @@ import (
 )
 
 //start starts the procedure
-func (rc *resec) start() {
+func (rc *resec) run() {
+	defer func() { rc.cleanup() }()
+
 	for {
 		select {
 		// got signal from the OS
@@ -201,8 +203,8 @@ func (rc *resec) parseMasterInfo(consulServiceInfo *consulapi.ServiceEntry) redi
 	return info
 }
 
-//stop stops the procedure
-func (rc *resec) stop() {
+// cleanup will clenup locks and similar internal state
+func (rc *resec) cleanup() {
 	rc.releaseConsulLock()
 
 	if rc.consul.serviceID != "" {

--- a/resec.go
+++ b/resec.go
@@ -2,23 +2,22 @@ package main
 
 import (
 	"log"
-	"os"
-	"os/signal"
-	"syscall"
 
 	consulapi "github.com/hashicorp/consul/api"
 )
 
 //start starts the procedure
 func (rc *resec) start() {
-	c := make(chan os.Signal, 1)
-	signal.Notify(c, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGINT)
-
 	for {
 		select {
 		// got signal from the OS
-		case <-c:
+		case <-rc.sigCh:
 			log.Printf("[INFO] Caught signal, stopping worker loop")
+			return
+
+		// got internal request to shut down
+		case <-rc.stopCh:
+			log.Printf("[INFO] Shutdown requested, stopping worker loop")
 			return
 
 		// got an update on redis replication status

